runtime/vm/classsupport.c

```diff
@@ -548,11 +548,12 @@ callFindLocallyDefinedClass(J9VMThread* vmThread, J9Module *j9module, U_8* class
 
 	if (NULL != dynamicLoadBuffers) {
 		 J9ClassPathEntry* classPathEntries = NULL;
+		 UDATA cpEntryCount = classLoader->classPathEntryCount;
```

Comment:
This doesn't really fix the problem, it just makes it less likely because the cpu can reorder the operations. There needs to be a read barrier after this line.

Reply:
Yes, this really does nothing - it's less likely by maybe a nanosecond.

Reply:
What is causing the problem is a new `cpEntryCount` and old `classPathEntries`. A new `classPathEntries` and old `cpEntryCount` is fine. If we want to add a memory barrier, I think it should be a write barrier in `addJarToSystemClassLoaderClassPathEntries()` which ensure `cpEntryCount` is updated after `classPathEntries`.

Reply:
Write barrier added

Reply:
I think the read barrier is also needed, the reads can still occur out of order.

Reply:
Yes.

Reply:
I'm still not convinced this will work - the old entries are being freed after assigning the new, so there's always a possibility of using freed memory. The only correct way to fix this is with a mutex.

Reply:
> so there's always a possibility of using freed memory. The only correct way to fix this is with a mutex.

Right, I missed that. Tried vm->classTableMutex. It will result in a deadlock. Seems that we need to add a new mutex.

Reply:
Both readers and writers will need to be protected, so you'll have to hold a mutex across the hook call, which is somewhere between dangerous and outright incorrect, depending on what the listeners are allowed to do.

Reply:
Currently only `hookFindSharedClass()` is registered with `J9HOOK_VM_FIND_LOCALLY_DEFINED_CLASS`. Nothing is registered with `J9HOOK_VM_FIND_LOCALLY_DEFINED_CLASS_FROM_FILESYSTEM`. 

I see we already pass `classLoader` through `TRIGGER_J9HOOK_VM_FIND_LOCALLY_DEFINED_CLASS`, how about acquiring and releasing the new mutex inside `hookFindSharedClass()`, and read the cp entries and count inside the mutex there.
The same for `findLocallyDefinedClass()`.

Reply:
That sounds like a safer solution. Probably best to remove the cpentry/count parameters from the hook entirely.

Reply:
> Probably best to remove the cpentry/count parameters from the hook entirely.

Unfortunately we are not able to remove the cpentry/count from the hook. The hook is used by all class loaders. For custom class loaders, the cp entry may not be classLoader->classPathEntries. It could be a different url specified by the class loader.

Reply:
Can you provide an example? I would call this change necessary as part of this work to avoid recurrence of the same issues.

Reply:
If nothing else, in the case where the values can't be trusted, 0s should be passed for the two parameters so if anyone does try to use them they will crash.

Reply:
> Can you provide an example? 

https://github.com/eclipse-openj9/openj9/blob/c101e0212fe158b0585ab75edab50d9cf99bdaf5/runtime/jcl/common/shared.c#L1290

The urlEntry is from the urlObj passed in from the JCL. Not classLoader->classPathEntries.

Reply:
OK, so perhaps the hook contract should be that if classPathEntries is NULL, you need to mutex and go to the classLoader to get the data, otherwise, trust the passed entries/count with no need to lock.